File: 3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int>> result;
        int target,last;
        if(num.size()<3) return result;
        sort(num.begin(),num.end());
        for(int w = 0 ; w < num.size() ; ++w){
            if(w>=1&&num[w] == num[w-1]) continue;
            target = -num[w];
            int it1 = w+1 , it2 = num.size()-1;
            while(it1<it2){
                if(num[it1]+num[it2]==target){
                    int sum = num[it1];
                    while((++it1)<it2&&num[it1]==sum);
                    --it1;
                    sum = num[it2];
                    while((--it2)>it1&&num[it2]==sum);
                    ++it2;
                    vector<int> temp{num[it1],num[it2],-target};
                    sort(temp.begin(),temp.end());
                    result.push_back(temp);
                    ++it1;
                    --it2;
                }
                else if(num[it1]+num[it2]<target) ++it1;
                else --it2;
            }
        }
        return result;
    }
};

File: java/3Sum.java
// first sort the array in O(nlogn)
// then,find all the cases by two point method in O(n^2)
// so, the solution's time complexity is O(n^2), space is O(n);
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>() ;
        Arrays.sort(nums) ;
        for (int i = 0 ; i < nums.length ; ++i) {
            if (i-1 >= 0 && nums[i] == nums[i-1]) continue;
            int target = nums[i] ;
            int low = i + 1, high = nums.length - 1;
            while (low < high) {
                if (nums[low] + nums[high] + target == 0) {
                    List<Integer> list = Arrays.asList(nums[i],nums[low],nums[high]);
                    result.add(list) ;
                    while (++low < high && nums[low-1] == nums[low]);
                    while (--high > low && nums[high+1] == nums[high]);
                }
                else if (nums[low] + nums[high] + target < 0) ++low;
                else --high;
            }
        }
        return result ;
    }
}
